zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+            None => None,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Return the most recently verified checkpoint's hash.
+    ///
+    /// If verification has not started yet, returns None. To verify the genesis
+    /// block, check that the parent block hash is all zeroes. (Bitcoin "null".)
+    ///
+    /// If verification has finished, returns the maximum checkpoint's hash.
+    fn get_previous_checkpoint_hash(&self) -> Option<BlockHeaderHash> {
+        match self.get_previous_checkpoint_height() {
+            // Every checkpoint height must have a hash
+            Some(height) => Some(*self.checkpoint_list.get(&height).unwrap()),
+            None => None,
+        }
+    }
+
+    /// Return the hash of the next checkpoint we want to verify.
+    ///
+    /// If verification has not started yet, returns the hash of block zero
+    /// (the genesis block). If verification has finished, returns None.
+    fn get_next_checkpoint_hash(&self) -> Option<BlockHeaderHash> {
+        match self.get_next_checkpoint_height() {
+            // Every checkpoint height must have a hash
+            Some(height) => Some(*self.checkpoint_list.get(&height).unwrap()),
+            None => None,
+        }
+    }
+
+    /// Return the height of the next checkpoint higher than `after_block_height`
+    /// in the checkpoint list. Ignores the current checkpoint range.
+    ///
+    /// If `after_block_height` is None, assume verification has not started yet,
+    /// and return zero (the genesis block).
+    ///
+    /// Returns None if there are no checkpoints. Also returns None if
+    /// `after_block_height` is greater than or equal to the maximum height.
+    fn find_descendant_checkpoint_height(
+        &self,
+        after_block_height: Option<BlockHeight>,
+    ) -> Option<BlockHeight> {
+        match after_block_height {
+            None => self.checkpoint_list.keys().cloned().next(),
+            Some(height) => self
+                .checkpoint_list
+                .range((Excluded(height), Unbounded))
+                .next()
+                .map(|(height, _)| *height),
+        }
+    }
+
+    /// Increase the current checkpoint height to `verified_block_height`,
+    /// if `verified_block_height` is the next checkpoint height.
+    fn update_current_checkpoint_height(&mut self, verified_block_height: BlockHeight) {
+        let previous_height = self.get_previous_checkpoint_height();
+        // Ignore blocks that are below the previous checkpoint.
+        //
+        // We ignore out-of-order verification, such as:
+        //  - the height is less than the previous checkpoint height, or
+        //  - the previous checkpoint height is the maximum height (checkpoint verifies are finished),
+        // because futures might not resolve in height order.
+        if let Some(previous) = previous_height {
+            if verified_block_height <= previous {
+                return;
+            }
+        }
```

Comment:
I think you can collapse these conditions:

```suggestion
        if previous_height >= Some(verified_block_height) {
                return;
        }
```

Reply:
Are you sure? How does `None` compare with `Some(_)`?

But I can collapse the statement after this one.

Reply:
`Ord` is derived for `Option`:
https://doc.rust-lang.org/src/core/option.rs.html#153

When derived, `Ord` creates an order based on enum declaration order:
https://doc.rust-lang.org/std/cmp/trait.Ord.html#derivable

So `Some > None`:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d039b60fe6be5d1daa09b6e53eba98bf

But I'm not sure I want to rely on hidden behaviour like that. Instead, I'll try to remove this comparison during the enum refinement in https://github.com/ZcashFoundation/zebra/pull/524#discussion_r447907746

Reply:
> But I can collapse the statement after this one.

Fixed in 018bff414adc012faecd6b77b953f04ebe566930.

Reply:
I rewrote this check as part of the broader rewrite.